clients/bazel-module-registry/src/funTest/kotlin/BazelModuleRegistryClientFunTest.kt

```diff
@@ -46,4 +46,28 @@ class BazelModuleRegistryClientFunTest : WordSpec({
             sourceInfo.integrity shouldBe "sha256-7t5x8oNxvzmqabRd4jsynTchQBbiBVJps7Xnz9QLWfU="
         }
     }
+
+    "CompositeBazelModuleRegistryService.kt.create" should {
+        "parse the 'registryFileHashes' property of the lock file (Bazel >= 7.2.0)" {
+            val expr1 = "https://raw.githubusercontent.com/bazelbuild/bazel-central-registry/main/modules/abseil-cpp/" +
+                "20230125.1/source.json"
+            val group1 = CompositeBazelModuleRegistryService.Companion.URL_REGEX.matchEntire(expr1)?.groups
+
+            group1?.get("server")
+                ?.value shouldBe "https://raw.githubusercontent.com/bazelbuild/bazel-central-registry/main/"
+            group1?.get("package")?.value shouldBe "abseil-cpp"
+
+            val expr2 = "https://bcr.bazel.build/modules/rules_proto/5.3.0-21.7/source.json"
+            val group2 = CompositeBazelModuleRegistryService.Companion.URL_REGEX.matchEntire(expr2)?.groups
+
+            group2?.get("server")?.value shouldBe "https://bcr.bazel.build/"
+            group2?.get("package")?.value shouldBe "rules_proto"
+
+            val expr3 = "https://bcr.bazel.build/modules/upb/0.0.0-20220923-a547704/source.json"
+            val group3 = CompositeBazelModuleRegistryService.Companion.URL_REGEX.matchEntire(expr3)?.groups
+
+            group3?.get("server")?.value shouldBe "https://bcr.bazel.build/"
+            group3?.get("package")?.value shouldBe "upb"
```

Comment:
This looks like a test for `CompositeBazelModuleRegistryService` rather than for `Bazel`.
Would the better be a unit test?

If the logic examined by this test can be easily covered with a functional test for `Bazel` , I'd strongly prefer such a fun test for bazel version 7.2.0 comparing against an expected resolveDevepencies() result.

Reply:
This test tests only a part of the `CompositeBazelModuleRegistryService`: the parsing of the `registryFileHashes`. While it _could_ be covered by a functional test as you mentioned, I see really a added benefit of adding a small test focusing on this specific aspect of the service: errors are way faster to identify.

In this sense, this is indeed a unit test. Could you recommend how to migrate it to a unit test?

Please note that the situation of test for Bazel is a bit confusing:

- there is this test, the `BazelModuleRegistryClientFunTest` which is not really a functional test as it test just bits of the Bazel package manager
- there is `BazelFunTest` which is a _real_ functional test that compares expected result
- there is `BazelTest` that test just a feature of the package manager.

Any advice how to refactor this would be appreciated.